manual/source/topic/arena.rst

```diff
@@ -1040,3 +1051,79 @@ Arena introspection and debugging
+
+.. _topic-arena-extension:
+
+Arena extension callbacks
+-------------------------
+
+There are situations in which the :term:`client program` needs to be
+informed about the chunks of address space that an :term:`arena` is
+managing. To support this, the MPS allows the client program to
+specify two callback functions when creating a :term:`virtual memory
+arena`: one function is called when the arena is *extended* (that is,
+when it acquires a new chunk of address space from the operating
+system), and the other when the arena is *contracted* (that is, when
+it returns a chunk of address space to the operating system).
+
+The use case that this feature is designed to support is debugging of
+dynamically generated code in 64-bit Windows. Microsoft's
+documentation for |RtlInstallFunctionTableCallback|_ says:
+
+    Function tables are used on 64-bit Windows to determine how to
+    unwind or walk the stack. These tables are usually generated by
+    the compiler and stored as part of the image. However,
+    applications must provide the function table for dynamically
+    generated code.
+
+An application may install a dynamic function table by calling
+|RtlInstallFunctionTableCallback|_, passing the region of memory in
+which the dynamically generated functions can be found, and may later
+delete the table by calling |RtlDeleteFunctionTable|_.
+
+.. |RtlDeleteFunctionTable| replace:: :c:func:`RtlDeleteFunctionTable`
+.. _RtlDeleteFunctionTable: https://docs.microsoft.com/en-gb/windows/win32/api/winnt/nf-winnt-rtldeletefunctiontable
+
+So if the client program is storing dynamically generated functions in
+MPS-managed memory, then it could define callback functions that
+install and delete the function table callback for the dynamically
+generated code, like this::
+
+    void arena_extended(mps_arena_t arena, void *base, size_t size)
+    {
+        RtlInstallFunctionTableCallback(...);
+    }
+
+    void arena_contracted(mps_arena_t arena, void *base, size_t size)
+    {
+        RtlDeleteFunctionTable(...);
+    }
+
+and then pass these two functions using :term:`keyword arguments` to
+:c:func:`mps_arena_create_k`::
+
+    MPS_ARGS_BEGIN(args) {
+        MPS_ARGS_ADD(args, MPS_KEY_ARENA_EXTENDED, (mps_fun_t)arena_extended);
+        MPS_ARGS_ADD(args, MPS_KEY_ARENA_CONTRACTED, (mps_fun_t)arena_contracted);
+        /* ... other keyword arguments ... */
+        res = mps_arena_create_k(&arena, mps_arena_class_vm(), args);
+    } MPS_ARGS_END(args);
+
+The callback functions receive three arguments: ``arena`` (the arena
+being extended or contracted), ``base`` (the base address of the chunk
+of address space that has just been acquired from, or is about to be
+returned to, the operating system), and ``size`` (the size of the
+chunk, in bytes). They must not call any function in the MPS, and must
+not access any memory managed by the MPS.
+
+.. note::
+
+    Arena extension callbacks are only supported by :term:`virtual
+    memory arenas`.
```

Comment:
@rptb1 this is possibly another case where https://github.com/Ravenbrook/mps/pull/223#discussion_r1197863623 applies,

Reply:
In the case of https://github.com/Ravenbrook/mps/pull/223#discussion_r1197863623 the problem is that the information does not *help* the developer.  They can't necessarily tell whether an address is in an AMC or AMCZ pool, and what effect this had on the return code.  That's a consequence of the poor design of the interface and abuse of result codes. None of that applies in this case.  The fact stated is useful and complete.